triton-constraint-circuit/src/lib.rs

```diff
@@ -775,6 +763,41 @@ impl<II: InputIndicator> ConstraintCircuitMonad<II> {
         (main_constraints, aux_constraints)
     }
 
+    fn substitute_node_and_produce_constraint(
+        multicircuit: &mut [Self],
+        info: DegreeLoweringInfo,
+        chosen_node_id: usize,
+        main_constraints_count: usize,
+        aux_constraints_count: usize,
```

Comment:
f9f83dc675016182b4f3bbe2a57f8fcbb9e8cdc6 Introduces type aliases `EvolvingMainConstraintsNumber` and `EvolvingAuxConstraintsNumber` for the second two counts.

Personally, I think this change degrades code quality. Having the commit as part of the PR is the fastest way forward, because it either earns your checkmark directly or else contributes to convincing you to reject it.

My position, in case you are interested:
 - Types used once are mis-used.
 - The function is called in one place only, not counting the test. It is conceptually part of that same function; the only reason why it is factored out is to test it in isolation.
 - The counts serve to derive indices of new columns and are thus inherently `usize`s. Declaring another type for these variables suggests a different context applies, but I don't think it does. The reader's attention is diverted towards finding this missing context.
 - Messing up the order of the arguments is still possible. (If due to too-weak introduced types, please show by example how it ought to be done.)
 - Why not declare a new type for every argument in every function interface? I do think that's the logical conclusion of your train of thought; if not I wonder where it stops.
 - All of the above being said, I do think that `NodeId` could be a useful type but a) that doesn't resolve the problem completely; and b) that refactor is out of the scope of this PR.

Reply:
[Type aliases](https://doc.rust-lang.org/book/ch19-04-advanced-types.html#creating-type-synonyms-with-type-aliases) don't do what you suggest they do in 66b8225 – that's [newtypes](https://doc.rust-lang.org/book/ch19-04-advanced-types.html#using-the-newtype-pattern-for-type-safety-and-abstraction).

Taking position:

> Types used once are mis-used.

I see help from the type system to write correct programs as beneficial in all cases.[^counter]

> The function is called in one place only […].

I don't see how this justifies bad design.

>  […] the […] reason why it is factored out is to test it in isolation.

Which is laudable and a decision I fully support.

> The counts […] are […] inherently `usize`s.

Consider:

```rust
let _ = PrintConfig::new(true, true, false);
let _ = PrintConfig::new(Sided::Single, Colored::Colored, PrintSpeed::Slow);
```

Even though (for arguments sake) all these enums have only two variants, and are thus inherently `bool`s, I strongly prefer the second version.[^better] In my opinion, the argument extends to numerical types.

> Why not declare a new type for every argument in every function interface?

The concrete contract of a function as well the number of arguments it takes make a huge difference. In `my_vec.remove(i)`, there's no room for screwing up the argument order.[^tech] The same holds for `my_vec.swap(i, j)`, and so having two arguments of the same type does not influence the method's usability. I don't have a hard-and-fast rule at hand as to where it is beneficial to introduce new types.

> I do think that `NodeId` could be a useful type

Sounds like a good idea. I agree with it being out of scope for this PR.

---

All that said, I think testability completely justifies factoring out the function. I also think that you have inherited design decisions (and, worse, accidental design) that would require a ton of work to improve.[^ind] We can write the argument sublist `usize, usize, usize` off as being part of overall improvable design, improving which is left as future work. One might argue that this constitutes building up technical debt,[^debt] but I think an equally valid argument is that factoring out the method only highlights existing technical debt.

In my opinion, introducing a type like `NumConstraints` slightly mitigates the issues I have with the same-type arguments. Fully resolving all pain points I see would probably require a more fundamental redesign.

[^better]: Other designs than a function `new()` taking many arguments should be considered were `PrintConfig` actually being added. The types' names would need a rework, too.
[^tech]: Technically, it's possible: `Vec::remove(i, &mut my_vec)`. However, (a) the type checker will stop you, and (b) why would you write that, apart from trying to prove a point, which, fair enough.
[^counter]: Yes, I'm looking forward to your counter argument after having made such an absolute claim.
[^ind]:  One strong indicator of this is that the two arguments in question are essentially internal tracking information. The function should not have to rely on externally supplied information for this.
[^debt]:  I'm sure you won't be surprised to hear that I'm not a fan of building up technical debt, and even less of doing it consciously.

Reply:
I modified 66b8225a (now 14a0572) to use newtypes instead of type aliases. Thanks for the reference. Let's proceed with this.

> I see help from the type system to write correct programs as beneficial in all cases.[1](https://github.com/TritonVM/triton-vm/pull/325#user-content-fn-counter-d762d723ea4b4f8dc30c9ddaf200bdf8)

You already read a few general counter-arguments in my bullet point list. In response to this particular phrasing: in the eyes of one person, "help from the type system to write correct programs" is "having to fight the type system to write anything at all in exchange for a marginal assurance of correctness" in the eyes of another. There are programming languages whose type systems are powerful enough to produce a formal computer-verifiable proof of programs' correctness. If rust's type system had this feature, would you recommend using it everywhere and always? The alternative would be to find an optimum on the tradeoff curve interpolating between (a) getting stuff done and (b) correctness.

> >  The function is called in one place only […].
> 
> I don't see how this justifies bad design.

To the extent that the partially elided sentence justifies deviating from your standards (I would obviously not use the B-word) the point is not so much that the function is used in a small number of places and therefore unimpactful; instead, the point made by the elided part is that the function is conceptually and semantically, inseparably part of the of the function that calls it. This relation makes the maintenance of a succinct and misuse-resistant interface moot at best and counter-productive at worst -- because if the algorithm changes then the "interface" will have to adapt.

> Consider:
> 
> ```rust
> let _ = PrintConfig::new(true, true, false);
> let _ = PrintConfig::new(Sided::Single, Colored::Colored, PrintSpeed::Slow);
> ```
> 
> Even though (for arguments sake) all these enums have only two variants, and are thus inherently bools, I strongly prefer the second version.[2](https://github.com/TritonVM/triton-vm/pull/325#user-content-fn-better-d762d723ea4b4f8dc30c9ddaf200bdf8) In my opinion, the argument extends to numerical types.

There is no natural mapping from `Single-Sided`/`Double-Sided` or `Colored`/`Monochrome` or `Slow`/`Fast` to `true`/`false`. You can define a mapping and call it canonical but it would still be an arbitrary choice that preserves no structure. So I disagree with the position that these enums "are inherently bools" and would even argue that the second line is preferable to the first.

To bolster my position that the counts in question *are* inherently `usize`s, consider the necessary boilerplate following the definition of `EvolvingMainConstraintNumber`:
```rust
struct EvolvingMainConstraintsNumber(usize);
impl From<EvolvingMainConstraintsNumber> for usize {
    fn from(value: EvolvingMainConstraintsNumber) -> Self {
        value.0
    }
}
```

In fact, the counts are not used *except* through this `from` method. The type is not used in any algebraic sense.

Maybe that's the point and the essence of our difference. I want types to semantically reflect the underlying maths; whereas you want types to constrain control flow and interfaces.

> In my_vec.remove(i), there's no room for screwing up the argument order.[3](https://github.com/TritonVM/triton-vm/pull/325#user-content-fn-tech-d762d723ea4b4f8dc30c9ddaf200bdf8) The same holds for my_vec.swap(i, j), and so having two arguments of the same type does not influence the method's usability.

There is no room for screwing up the argument order (footnote notwithstanding) but there is room for making the wrong selection from the variables of type `usize` that live in scope. In this case, binding the interface through custom defined types gives the false impression of correctness, which is worse than no impression at all.